benchmark_decode_test.go

```diff
@@ -0,0 +1,283 @@
+package redis
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-redis/redis/v8/internal/proto"
+	"net"
+	"testing"
+	"time"
+)
+
+var ctx = context.TODO()
+
+type BenchConn struct {
+	buff []byte
+}
+
+func (c *BenchConn) Read(b []byte) (n int, err error)   { return copy(b, c.buff), nil }
+func (c *BenchConn) Write(b []byte) (n int, err error)  { return len(b), nil }
+func (c *BenchConn) Close() error                       { return nil }
+func (c *BenchConn) LocalAddr() net.Addr                { return nil }
+func (c *BenchConn) RemoteAddr() net.Addr               { return nil }
+func (c *BenchConn) SetDeadline(_ time.Time) error      { return nil }
+func (c *BenchConn) SetReadDeadline(_ time.Time) error  { return nil }
+func (c *BenchConn) SetWriteDeadline(_ time.Time) error { return nil }
+
+func NewDecodeClient() *RespDecode {
+	conn := &BenchConn{}
+	client := NewClient(&Options{
+		PoolSize: 128,
+		Dialer: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return conn, nil
+		},
+	})
+	return &RespDecode{
+		name:   "Client",
+		conn:   conn,
+		client: client,
+	}
+}
+
+func NewDecodeClusterClient() *RespDecode {
+	conn := &BenchConn{}
+	client := NewClusterClient(&ClusterOptions{
+		PoolSize: 128,
+		Addrs:    []string{"127.0.0.1:6379"},
+		Dialer: func(ctx context.Context, network, addr string) (net.Conn, error) {
+			return conn, nil
+		},
+		ClusterSlots: func(_ context.Context) ([]ClusterSlot, error) {
+			return []ClusterSlot{
+				{
+					Start: 0,
+					End:   16383,
+					Nodes: []ClusterNode{{Addr: "127.0.0.1:6379"}},
+				},
+			}, nil
+		},
+	})
+	// init command.
+	tmpClient := NewClient(&Options{Addr: ":6379"})
+	cmdCache, err := tmpClient.Command(ctx).Result()
+	_ = tmpClient.Close()
+	client.cmdsInfoCache = newCmdsInfoCache(func(_ context.Context) (map[string]*CommandInfo, error) {
+		return cmdCache, err
+	})
+
+	return &RespDecode{
+		name:   "Cluster",
+		conn:   conn,
+		client: client,
+	}
+}
+
+type RespDecode struct {
+	name   string
+	conn   *BenchConn
+	client Cmdable
+}
```

Comment:
What do you think about tweaking API a bit

```go
type ClientStub struct {
    Cmdable
    conn *ConnStub
}

// Or StubResponse?
func (ClientStub) SetResponse(b []byte) {}

stub.SetResponse([]byte("something"))
stub.Ping(ctx)
```

Reply:
It looks good, I have made changes and supported large data volume testing